Cython/Compiler/ExprNodes.py

```diff
@@ -3119,19 +3119,46 @@ def generate_next_sequence_item(self, test_name, result_name, code):
             inc_dec = '--'
         else:
             inc_dec = '++'
-        code.putln("#if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS")
-        code.putln(
-            "%s = Py%s_GET_ITEM(%s, %s); __Pyx_INCREF(%s); %s%s; %s" % (
-                result_name,
-                test_name,
-                self.py_result(),
-                self.counter_cname,
-                result_name,
-                self.counter_cname,
-                inc_dec,
-                # use the error label to avoid C compiler warnings if we only use it below
-                code.error_goto_if_neg('0', self.pos)
-                ))
+        if test_name == 'List':
+            code.putln("#if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS && !CYTHON_AVOID_THREAD_UNSAFE_BORROWED_REFS")
+            code.putln(
+                "%s = Py%s_GET_ITEM(%s, %s); __Pyx_INCREF(%s); %s%s; %s" % (
+                    result_name,
+                    test_name,
+                    self.py_result(),
+                    self.counter_cname,
+                    result_name,
+                    self.counter_cname,
+                    inc_dec,
+                    # use the error label to avoid C compiler warnings if we only use it below
+                    code.error_goto_if_neg('0', self.pos)
+                    ))
+            code.putln("#elif CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS")
+            code.putln(
+                "%s = Py%s_GetItemRef(%s, %s); %s%s; %s" % (
```

Comment:
I think this'll need a `__Pyx_GOTREF(%s)` (in place of `__Pyx_INCREF` in the others) to pass our tests.

That's a no-op in a regular build, but when running with our "refnanny" tool it helps track the new reference.